crates/http-server/src/middleware/monitoring.rs

```diff
@@ -0,0 +1,64 @@
+use axum::{body::Body, http::Request, response::Response};
+use futures_util::future::BoxFuture;
+use std::task::{Context, Poll};
+use std::time::Instant;
+use tower::{Layer, Service};
+
+#[derive(Clone)]
+pub struct MonitoringLayer;
+
+impl<S> Layer<S> for MonitoringLayer {
+    type Service = MonitoringMiddleware<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        MonitoringMiddleware { inner }
+    }
+}
+
+#[derive(Clone)]
+pub struct MonitoringMiddleware<S> {
+    inner: S,
+}
+struct RequestInfo {
+    method: String,
+    path: String,
+}
+
+impl<S> Service<Request<Body>> for MonitoringMiddleware<S>
+where
+    S: Service<Request<Body>, Response = Response> + Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request<Body>) -> Self::Future {
+        let start_time = Instant::now();
+        let request_info = RequestInfo {
+            method: request.method().to_string(),
+            path: request.uri().path().to_string(),
+        };
```

Comment:
I think we can use `request.method().to_string()` and `request.uri().path().to_string()` straight!
No need to allocate `request_info` to drop it just after L53

https://github.com/http-server-rs/http-server/pull/326/files#diff-11aa8b5243fe114bb1329d03faa4979762046503a2fd2493683f1bdeeed12adbR53-R59

Reply:
This option is not possible straight!

Reply:
I wonder why? 🤔

In the docs I can see that `uri` and `method` borrow `self`:

https://docs.rs/http/0.2.9/http/request/struct.Request.html#method.method

https://docs.rs/http/0.2.9/http/request/struct.Request.html#method.uri

So it shouldn't be moving. Can you share the compile error please?

Reply:
Another improvement I see is saving the call to `to_string`, given that both:

- Uri: https://docs.rs/http/0.2.9/http/uri/struct.Uri.html#impl-Display-for-Uri
- Method: https://docs.rs/http/0.2.9/http/method/struct.Method.html#impl-Display-for-Method

Implement display.

Reply:
I think wrapping the request in an ARC can be an option like this:
https://github.com/http-server-rs/http-server/blob/main/src/server/middleware/logger.rs
But doing it, I think I'm going outside the scope of the issue.